sphinx_immaterial/code_annotations.py

```diff
@@ -0,0 +1,79 @@
+from typing import List
+from docutils import nodes
+from sphinx.application import Sphinx
+from sphinx.util.docutils import SphinxDirective
+from sphinx.writers.html import HTMLTranslator
+from sphinx.util.logging import getLogger
+from . import html_translator_mixin
+
+LOGGER = getLogger(__name__)
+
+
+class annotations_list(nodes.container):
+    pass
+
+
+class CodeAnnotations(SphinxDirective):
+    """Take a list of annotations and move them into a previous node's parent."""
+
+    has_content = True
+
+    def run(self) -> List[nodes.Node]:
+        """Run the directive."""
+        self.assert_has_content()
+
+        div = annotations_list("")
+        self.state.nested_parse(self.content, self.content_offset, div)
+        self.set_source_info(div)
+        if not isinstance(div.children[0], nodes.enumerated_list):
+            LOGGER.error(
+                "The code-annotations directive only accepts an enumerated list.",
+                location=div,
+            )
+        return [div]
+
+
+def visit_annotations_list(self: HTMLTranslator, node: annotations_list):
+    # the content should have been moved into the literal block before calling this
+    raise nodes.SkipNode()  # see visit_literal_block() override
+
+
+@html_translator_mixin.override
+def visit_literal_block(
+    self: html_translator_mixin.HTMLTranslatorMixin,
+    node: nodes.literal_block,
+    super_func: html_translator_mixin.BaseVisitCallback[nodes.literal_block],
+):
+    #  we need to be sure that the annotated_list is the next node
+    next_element = node.next_node(nodes.Element, siblings=True, descend=False)
```

Comment:
Move `next_element` definition below following `if` block so that it doesn't need to be duplicated

Reply:
I'm not sure I understand this request. If I move this definition below the
```py
if isinstance(parent.children[0], nodes.caption)
```
block then it would break captioned code-blocks because it overwrites the purpose of the if block.

The `next_element` definition isn't exactly duplicated because the if block takes the `next_node()` of the parent (due to the presence of the caption's extra div parent).

Reply:
I removed the unnecessary `parent` definition (really an artifact from my debugging `print` statements).

Reply:
You could simplify this to:

```python
annotation_parent = node
if isinstance(node.parent.children[0], nodes.caption):
  annotation_parent = node.parent
next_element = annotation_parent.next_node(nodes.Element, siblings=True, descend=False)
```